pkg/controllers/subnet/subnet_controller.go

```diff
@@ -242,5 +244,58 @@ func (r *SubnetReconciler) Start(mgr ctrl.Manager) error {
 	if err != nil {
 		return err
 	}
+	go r.GarbageCollector(make(chan bool), servicecommon.GCInterval)
 	return nil
 }
+
+func (r *SubnetReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("subnet garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		nsxSubnetList := r.Service.ListSubnet()
+		if len(nsxSubnetList) == 0 {
+			continue
+		}
+
+		crdSubnetList := &v1alpha1.SubnetList{}
+		err := r.Client.List(ctx, crdSubnetList)
+		if err != nil {
+			log.Error(err, "failed to list subnet CR")
+			continue
+		}
+
+		crdSubnetIDs := sets.NewString()
+		for _, sr := range crdSubnetList.Items {
+			crdSubnetIDs.Insert(string(sr.UID))
+		}
+
+		crdSubnetSetIDs := r.Service.ListSubnetSetID(ctx)
+		if len(crdSubnetSetIDs) == 0 {
+			continue
+		}
+
+		for _, elem := range nsxSubnetList {
```

Comment:
we do not need to iterate all the nsxSubnetList again, since the above already checked all the NSX subnets created for Subnet CRs. 
Could we remove those NSX subnets from nsxSubnetList if found they have corresponding Subnet CR?
Then in the left NSX Subnet list, check they have corresponding SubnetSet CR.  
Instead of checking NSX Subnet tags, could we use the store SubnetIndexFunc?

Reply:
split the function into subnet_controller/subnetset_controller gc